### DISCUSSION TITLE
Conversion failed when converting character string to smalldatetime data type.

### DISCUSSION THREAD

**Discussion by pedro-fiuza** [balta-io/forum#166] [Tira Dúvidas] [answered]
Boa noite @andrebaltieri,

Seguindo as instrucoes do HasDefaultValue para o datatype SMALLDATETIME ao tentar inserir a entidade no database ocorre o seguinte error: Conversion failed when converting character string to smalldatetime data type.

Mapping:

```
 builder.Property(x => x.ExpireDate)
           .IsRequired()
           .HasColumnName("ExpireDate")
           .HasColumnType("SMALLDATETIME")
           .HasDefaultValue(DateTime.Now.ToUniversalTime());
```

**Comment by andrebaltieri** [comment, marked as answer]
Bom dia, @pedro-fiuza, qual string você está tentando inserir como `SMALLDATETIME`? Qual valor?

**Reply by pedro-fiuza** [reply]
Fala @andrebaltieri, bom dia!

Eu não estou manipulando os campos de data, configurei eles como nullable e estou inserindo a entidade para que o HasDefaultValue faça o seu trabalho.

![image](https://user-images.githubusercontent.com/33269752/208116029-65e72001-415e-4075-995a-17a3ffd90f45.png)

No erro que esta ocorrendo, tentei entrar na exceção pra tentar visualizar o valor que esta pra inserção, mas não consegui, porem seguindo a logica, ele deve estar tentando inserir o valor configurado no HasDefaultValue, tendo em vista que os valores estão nulos na entidade (CreatedAt, UpdatedAt).

**Reply by pedro-fiuza** [reply]
Update:

@andrebaltieri, a unica forma que eu consegui fazer acontecer aqui foi retirando o DefaultValue e inicializando os campos com a data.
`
public DateTime? CreatedAt { get; set; } = DateTime.UtcNow;`

Ao fazer isso, consegui fazer a insercao corretamente, mas ainda nao consegui entender pq o HasDefaultValue nao esta funcionando como esperado.

**Reply by DiegoOliver94** [reply]
@pedro-fiuza boa tarde, tudo bem?

estava com o mesmo problema e realizando algumas simulações consegui adicionar no banco de dados nos cenarios abaixo:

 builder.Property(x => x.ExpireDate)
           .IsRequired()
           .HasColumnName("ExpireDate")
           .HasDefaultValue(DateTime.Now.ToUniversalTime());

aplicando através do código removendo a propriedade do tipo de coluna (   .HasColumnType("SMALLDATETIME")), ele aplicou e salvou normalmente.

e usando o GETDATE(); conforme o curso, neste caso explicitando o tipo de coluna:

 builder.Property(x => x.ExpireDate)
           .IsRequired()
           .HasColumnName("ExpireDate")
           .HasColumnType("SMALLDATETIME")
           .HasDefaultValuesql("GETDATE()");


das formas acima não apresentou erro ao salvar, espero ter ajudado
